### PR TITLE
Bluetooth: Host: add unregister connection callback function

### PR DESCRIPTION
**Pull request by zhjseu**
[Description]
tests: shell: Restart bt will register the same connection callback twice.
Callback next node point to itself, when link established callback function
loop infinitely.
[Fix]
Unregister the previous callback to avoid register repeatedly.
[Test]
After bt init/disable times, create connection successfully.

### CODE REVIEW

File: subsys/bluetooth/host/conn.c
Comment (Thalley) [inline subsys/bluetooth/host/conn.c:2356]: If `cb` is never found in the callback list, should we return an error?

Comment (zhjseu) [reply subsys/bluetooth/host/conn.c:2356]: HI Thalley, 
I think it's not need to return an error.
1. this is use to remove a callback, not to find a callback, if the callback not register before, it's ok.
2. It is better to have a uniform function format, like bt_le_scan_cb_unregister has no return.
thanks.

Comment (Thalley) [reply subsys/bluetooth/host/conn.c:2356]: Hmm, I think `bt_le_scan_cb_unregister` might be an outlier here, as `bt_conn_auth_info_cb_unregister`, `bt_gatt_service_unregister`, `bt_iso_server_unregister`, `bt_has_preset_unregister`, `bt_pacs_cap_unregister` and `bt_ots_client_unregister` all return an error if the thing being unregistered was not registered first. 

I think it would make sense to return -ENOENT or -EALREADY if the provided call is currently not registered. @jhedberg what do you think?

Comment (zhjseu) [reply subsys/bluetooth/host/conn.c]: HI @Thalley @jhedberg , I submit the code again, please help to review, thanks.

Comment (zhjseu) [reply subsys/bluetooth/host/conn.c]: Hi @Thalley @jhedberg , I submit the code again, please help to review, thanks.

File: include/zephyr/bluetooth/conn.h
Comment (Thalley) [inline include/zephyr/bluetooth/conn.h:1072]:
```suggestion
 * @retval 0 Success
 * @retval -EINVAL If @p cb is NULL
 * @retval -ENOENT if @p was not registered
```

File: include/zephyr/bluetooth/conn.h
Comment (Thalley) [inline include/zephyr/bluetooth/conn.h:1071]:
```suggestion
/** 
 * @brief Unregister connection callbacks.
 *
 * Unregister the state of connections callbacks.
 *
 * @param cb Callback struct point to memory that remains valid.
 *
```

File: subsys/bluetooth/host/conn.c
Comment (Thalley) [inline subsys/bluetooth/host/conn.c:2361]:
```suggestion
		}
		
		previous_callback = previous_callback->_next;
```

File: include/zephyr/bluetooth/conn.h
Comment (Thalley) [inline include/zephyr/bluetooth/conn.h:1074]: I think we should expand at least one of our tests (possibly babblesim (bsim) tests, or a unit test) to run and test this functionality.

**Review by Thalley — CHANGES_REQUESTED** [review]
Functionality looks good. 

Some documentation should be updated, and a code style comment. 

Finally, since this is adding a new function in the public API, we should test that it works as intended in CI.

File: include/zephyr/bluetooth/conn.h
Comment (Thalley) [inline include/zephyr/bluetooth/conn.h:1075]:
```suggestion
 * @retval -ENOENT if @p cb was not registered
```

**Review by Thalley — DISMISSED** [review]

**Review by Thalley — COMMENTED** [review]
Changes LGTM to me, but I do still think we should add a test of this before merging

**Comment by zhjseu** [comment]
Hello @Thalley , what should I do next?

**Comment by theob-pro** [comment]
> Hello @Thalley , what should I do next?

Hi @zhjseu, I think the next thing to do would be to add a test (bsim or unit test) to ensure the new API work as intended, as @Thalley suggested it.

**Comment by zhjseu** [comment]
> > Hello @Thalley , what should I do next?
> 
> Hi @zhjseu, I think the next thing to do would be to add a test (bsim or unit test) to ensure the new API work as intended, as @Thalley suggested it.

HI @theob-pro , 
I don't know how to add a bsim or unit test, but to make sure this API work or not, can do the follow test,
bt init -> bt disable -> bt init -> connected remote device
> if connect remote device success, this API work.
> if connect fail , this API not work.
thanks.

**Comment by Thalley** [comment]
> > > Hello @Thalley , what should I do next?
> > 
> > 
> > Hi @zhjseu, I think the next thing to do would be to add a test (bsim or unit test) to ensure the new API work as intended, as @Thalley suggested it.
> 
> HI @theob-pro , I don't know how to add a bsim or unit test, but to make sure this API work or not, can do the follow test, bt init -> bt disable -> bt init -> connect remote device, -> If connect remote device success, this API work. -> If connect fail , this API not work. thanks.

The thing we are requesting is that is it being properly tested by CI, so that we can be sure it works for more than just the shell and that it continues to work if there are other changes. 

It isn't too difficult to get started on babblesim (or even unit tests), so be don't afraid to look into it :)

**Comment by dleach02** [comment]
@Thalley, @theob-pro can you point @zhjseu to where you expect the test cases to exist in the test tree?

**Comment by zhjseu** [comment]
> > @Thalley, @theob-pro can you point/guide @zhjseu to where you expect the test cases to exist in the test tree?
> 
> I would create a new test under https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/bluetooth/host/misc that does the following:
> 
> 1. Register callbacks
> 2. Create a connection and verify that callback is called
> 3. Disconnect the connection
> 4. Unregister callbacks
> 5. Create a new connection and verify that the callback is not called

@Thalley, Thanks for your reply,
Are you going to add this test process or  shall I do it ？
I never add bsim test case before, is that call the register conn callback and unregister conn callback function in the follow file?
`/tests/bsim/bluetooth/host/misc/conn_stress/central/src/main.c`
To verify code added no compile error, could you tell me which script to run to build the bsim image?
Thanks!

**Comment by Thalley** [comment]
> > > @Thalley, @theob-pro can you point/guide @zhjseu to where you expect the test cases to exist in the test tree?
> > 
> > 
> > I would create a new test under https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/bluetooth/host/misc that does the following:
> > 
> > 1. Register callbacks
> > 2. Create a connection and verify that callback is called
> > 3. Disconnect the connection
> > 4. Unregister callbacks
> > 5. Create a new connection and verify that the callback is not called
> 
> @Thalley, Thanks for your reply, Are you going to add this test process or shall I do it ？ 

I think you should do it if you are up for it :) 

> I never add bsim test case before, is that call the register conn callback and unregister conn callback function in the follow file? `/tests/bsim/bluetooth/host/misc/conn_stress/central/src/main.c` 

I would create a new directory in https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/bluetooth/host/misc and create a new test from scratch (using e.g. https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/bsim/bluetooth/host/misc/disable as inspiration). That way we can have a specific test for this. 

> To verify code added no compile error, could you tell me which script to run to build the bsim image? Thanks!

You simply compile it as you would with any other application, but using the `nrf52_bsim` board. You can then a script similar to https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bsim/bluetooth/host/misc/disable/tests_scripts/disable_with_gatt.sh to run and verify the test.

**Comment by zhaynxp** [comment]
Hi @Thalley , reviewers, 

After setup BabbleSim environment in my local PC, compile  (west build -p always -b nrf52_bsim -d xxx xxx) still failed says: 
 _CMake Error at C:/home/nxf57794/Zephyr_projects/zephyr/arch/posix/CMakeLists.txt:4 (message):
  The POSIX architecture only works on Linux.  If on Windows or macOS
  consider using a virtual machine to run a Linux guest._

Can't Windows PC be used to compile and run the app target to `nrf52_bsim` ? Currently we only have Zephyr env in Windows PC to  facilitate development. 
Could you help to guide here? 
Thanks a lot!

**Comment by jori-nordic** [comment]
@zhaynxp yes bsim only works on linux. You can set up a virtual machine w/ ubuntu 22.04 and it should work in that environment.

**Comment by zhaynxp** [comment]
> @zhaynxp yes bsim only works on linux. You can set up a virtual machine w/ ubuntu 22.04 and it should work in that environment.

Hi @jori-nordic , thanks for the info. More one question is what if the bsim test pass on our local linux setup? any material or evidence needed to submit to this ticket for clarification? Could you please share next step？

**Comment by jori-nordic** [comment]
If the new test passes on your local setup, the next step is to add it to the pull request introducing the change.
E.g. like I did here: I identified a bug, made a test that reproduces it by failing, and then added the fix that makes the test pass again: https://github.com/zephyrproject-rtos/zephyr/pull/67531/files

From this PR you can see what you need to add a new test.

File: tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:172]: you should add a conn-foreach to find out if you are really connected here. In addition to the flag not being set

**Review by jori-nordic — CHANGES_REQUESTED** [review]

File: tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c
Comment (zhaynxp) [reply tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:172]: thanks Jori, have add below print in `device_found`() to indicate that connection did happen:
`printk("%s: connected to found device\n", __func__);  Please check the pass log:
`
image

is it ok in this case?

Comment (jori-nordic) [reply tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:172]: that's good for debugging, but the script needs to actually fail (i.e. return something else than 0) if that doesn't happen.
You need a `FAIL()` call in the case you're not connected. Else the CI is never going to fail, someone will have to inspect the logs.

Comment (zhaynxp) [reply tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:202]: good point, have added the fail logic, pls help to review again, thanks a lot!

File: tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c
Comment (jori-nordic) [inline tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:112]: since we're only creating LE connections, you could move this to L105.

**Review by jori-nordic — DISMISSED** [review]

File: tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c
Comment (jori-nordic) [reply tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c:202]: looks good now :+1:

File: include/zephyr/bluetooth/conn.h
Comment (jori-nordic) [reply include/zephyr/bluetooth/conn.h:1077]: They added a test, could you re-review?

File: tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c
Comment (zhaynxp) [reply tests/bsim/bluetooth/host/misc/unregister_conn_cb/src/main.c]: noted, resolved this, thanks a lot!

**Comment by zhaynxp** [comment]
Dear Reviewers, 

There are still 2 checks failed to this PR, one is bsim-test failure as below:
image

the other is compliance check failure as below:
image

Is there anything we need to do on above failures? Please kindly let us know if anything pending on our side,  really appreciated and many thanks!

**Comment by jori-nordic** [comment]
@zhaynxp seems the permissions are not set properly on the test you just made.
A `chmod +x unregister_conn_cb.sh` and `git commit --amend` should do the trick.

As for the other error, the workflow used by zephyr PRs is to always force-push, and rebase on `origin/main` if necessary. Basically never do a merge.

You can probably resolve that by doing a `git rebase -i origin/main`. In the interactive rebase, you can then remove the merge commit (if it shows up at all).

**Comment by zhaynxp** [comment]
> @zhaynxp seems the permissions are not set properly on the test you just made. A `chmod +x unregister_conn_cb.sh` should do the trick.
> 
> As for the other error, the workflow used by zephyr PRs is to always force-push, and rebase on `origin/main` if necessary. Basically never do a merge or small edition commits in response to comments.
> 
> You can probably resolve that by doing a `git rebase -i origin/main`. In the interactive rebase, you can then remove the merge commit (if it shows up at all).

Hello @jori-nordic , all, thanks for the suggestions. After above 2 problems solved, a new error occured:
image
also found the target_prj_conf was not being built/compiled before running test, we've shared the _compile.sh in this PR, but don't understand how and when the bsim-test check will compile the target_prj_conf, could you please guide? Many thanks!

**Comment by jori-nordic** [comment]
you should add it below here: https://github.com/zephyrproject-rtos/zephyr/blob/8534767aed7d411f3081b02ec271b253a3d4fde4/tests/bsim/bluetooth/host/compile.sh#L77

**Comment by zhaynxp** [comment]
Hi @jori-nordic @Thalley , would you please approve this PR if no more changes are needed? 
Also the bsim online check always fail, it seems to be network issue:
image
image
image

the bsim run on my local VM with below command would report failure: 
`$(west topdir)/zephyr/tests/bsim/bluetooth/host/compile.sh && SEARCH_PATH=$(west topdir)/zephyr/tests/bsim/bluetooth/host/misc/unregister_conn_cb $(west topdir)/zephyr/tests/bsim/run_parallel.sh`
image
but the target_prj_conf was built in target folder and can pass if manually run ./unregister_conn_cb.sh
Many thanks here!

**Review by jori-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]
I'm approving since considering the current callback struct definition, this is the only way to do it and the PR has been open for long enough without me pointing the following out: would it perhaps have made more sense to just convert the callback list to `sys_slist_t` and use `sys_snode_t` in the callback struct instead of a custom linked list pointer? Then we don't have to do this reimplementation of the linked list manipulation code. Anyway, can be done as a follow-up effort as far as I'm concerned.

**Review by Thalley — APPROVED** [review]

**Pull request merged**
